jepsen/src/jepsen/core.clj

```diff
@@ -138,72 +138,122 @@
-
-                    ; Sanity checks
-                    (let [t (:type completion)]
-                      (assert (or (= t :ok)
-                                  (= t :fail)
-                                  (= t :info))
-                              (str "Expected client/invoke! to return a map with :type :ok, :fail, or :info, but received " (pr-str completion) " instead")))
-                    (assert (= (:process op) (:process completion)))
-                    (assert (= (:f op)       (:f completion)))
-
-                    ; Log completion
-                    (conj-op! test completion)
-
-                    (if (or (knossos/ok? completion) (knossos/fail? completion))
-                      ; The process is now free to attempt another execution.
-                      process
-                      ; Process hung; move on
-                      (+ process (:concurrency test))))
-
-                  (catch Throwable t
-                    ; At this point all bets are off. If the client or network
-                    ; or DB crashed before doing anything; this operation won't
-                    ; be a part of the history. On the other hand, the DB may
-                    ; have applied this operation and we *don't know* about it;
-                    ; e.g.  because of timeout.
-                    ;
-                    ; This process is effectively hung; it can not initiate a
-                    ; new operation without violating the single-threaded
-                    ; process constraint. We cycle to a new process identifier,
-                    ; and leave the invocation uncompleted in the history.
-                    (conj-op! test (assoc op
-                                          :type :info
-                                          :time  (relative-time-nanos)
-                                          :error (str "indeterminate: "
-                                                      (if (.getCause t)
-                                                        (.. t getCause
-                                                            getMessage)
-                                                        (.getMessage t)))))
-                    (warn t "Process" process "indeterminate")
-                    (+ process (:concurrency test))))))))
-        (info "Worker" process "done")))))
+        (info "Worker" process "with node" node "starting")
+
+        ; Differentiate our base client for this process
+        (let [client (client/open-compat! (:client test) test node)
+              _      (assert client (str "Expected a client, but got " (pr-str client)
+                                         " instead."))
+
+              ; Wait to begin ops until all clients are set up
+              _      (.await setup-barrier)
+
+              ; Consume operations from the generator, returns most recent client when done
+              client (loop [[process client] [process client]]
```

Comment:
We're close here, but this section can still throw in a few ways--notably, the assert will throw an AssertionError and bail out immediately without closing the client. What about... wrapping the whole thing in a `try`, and using (finally (.await setup-barrier ...)) to ensure cleanup happens regardless of exceptions?

Reply:
(I recognize this is A Lot Of Nesting (TM)--feel free to break up into more fns as you see fit!)

Reply:
Thank you for catching this! I believe this has been addressed in the updated mechanics.Took a bit of fiddling, but I got the clj compiler to be happy with the arrangement of try/catch/finally. I feel like a `with-client` macro could address some of the repetition here, but I'm not sure how much the indirection would win us in the single case. Additional feedback on how to improve this section would be appreciated.